rustls-symcrypt/src/signer.rs

```diff
@@ -1,22 +1,389 @@
+                )
+                .map_err(|_| Error::General("Failed to set ECDSA key from SEC1".into()))?
+            }
+            PrivateKeyDer::Pkcs8(pkcs8) => {
+                // Extract DER-encoded private key blob for PKCS#8
+                let private_key_blob = pkcs8.secret_pkcs8_der();
+                // Parse the DER-encoded private key
+                let private_key_info = PrivateKeyInfo::from_der(private_key_blob);
+                // Parse the PKCS#8 DER-encoded EC private key
+                let private_key = EcPrivateKey::from_der(&private_key_info.unwrap().private_key)
+                    .map_err(|_| Error::General("Failed to parse PKCS#8 DER".into()))?;
+
+                // Use EcPrivateKey's private_key to set up the ECDSA key
+                EcKey::set_key_pair(
+                    curve_type,
+                    &private_key.private_key,
+                    None,
+                    EcKeyUsage::EcDsa,
+                )
+                .map_err(|_| Error::General("Failed to set ECDSA key from PKCS#8".into()))?
+            }
+            _ => {
+                return Err(Error::General(
+                    "Invalid key format: must be PKCS#1 or PKCS#8".into(),
+                ))
+            }
+        };
+
+        // Return the ECDSASigningKey with Arc-wrapped key_pair and scheme
+        Ok(Self {
+            key: Arc::new(key),
+            scheme,
+        })
+    }
+}
+
+impl SigningKey for EcdsaSigningKey {
+    fn choose_scheme(&self, offered: &[SignatureScheme]) -> Option<Box<dyn Signer>> {
+        if offered.contains(&self.scheme) {
+            Some(Box::new(EcdsaSigner {
+                key: Arc::clone(&self.key),
+                scheme: self.scheme,
+            }))
+        } else {
+            None
+        }
+    }
+
+    fn algorithm(&self) -> SignatureAlgorithm {
+        SignatureAlgorithm::ECDSA
+    }
+}
+
+#[derive(Debug)]
+struct EcdsaSigner {
+    key: Arc<EcKey>,
+    scheme: SignatureScheme,
+}
+
+
+impl Signer for EcdsaSigner {
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
+        // Step 1: Hash the message based on the scheme
+        let hash_value = match self.scheme {
+            SignatureScheme::ECDSA_NISTP256_SHA256 => sha256(message).to_vec(),
+            SignatureScheme::ECDSA_NISTP384_SHA384 => sha384(message).to_vec(),
+            SignatureScheme::ECDSA_NISTP521_SHA512 => sha512(message).to_vec(),
+            _ => return Err(Error::General("unsupported ECDSA signature scheme".into())),
+        };
+
+        // Step 2: Sign the hashed message
+        let signature = self
+            .key
+            .ecdsa_sign(&hash_value)
+            .map_err(|e| Error::General(format!("failed to sign message: {}", e)))?;
+
+        // Step 3: Split the signature into r and s components
+        let (r, s) = signature.split_at(signature.len() / 2);
+
+        // Step 4: Create an RsaPublicKey structure with r and s
```

Comment:
>Step 4: Create an RsaPublicKey structure with r and s [](http://example.com/codeflow?start=11&length=53)

Add comments like: ECSignatureData is encoded as sequence of two integers. RsaPublicKey is also encoded as sequence of two integers. Will use RsaPublicKey to enode where modulus contains r and public_exponent contains s